cpp/src/arrow/util/spaced.h

```diff
@@ -0,0 +1,266 @@
+  int64_t idx_valid_bits = valid_bits_offset + idx_spaced;
+
+  const int64_t offset_suffix_back = (idx_valid_bits + 1) % kBatchSize;
+  // First handle back suffix
+  for (int64_t i = 0; (i < offset_suffix_back) && (idx_decode < idx_spaced); i++) {
+    if (BitUtil::GetBit(valid_bits, idx_valid_bits)) {
+      buffer[idx_spaced] = buffer[idx_decode];
+      idx_decode--;
+    }
+    idx_spaced--;
+    idx_valid_bits--;
+  }
+
+  // The parts can fill into batches, index from back.
+  int64_t idx_valid_bytes = BitUtil::BytesForBits(idx_valid_bits + 1) - 1;
+  uint8_t decode_count;
+  static const __m512i zero = _mm512_set_epi64(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0);
+  while (idx_decode < idx_spaced && (idx_spaced + 1) >= kBatchSize) {
+    // count decode numbers of one batch and step the index
+    decode_count = BitUtil::kBytePopcount[valid_bits[idx_valid_bytes]];
+    idx_valid_bytes--;
+    if (kBatchValidBytes > 1) {
+      decode_count += BitUtil::kBytePopcount[valid_bits[idx_valid_bytes]];
+      idx_valid_bytes--;
+    }
+    idx_decode -= decode_count;
+    idx_spaced -= kBatchSize;
+    idx_valid_bits -= kBatchSize;
+
+    // pack the data
+    if (decode_count > 0) {
+      __m512i src = _mm512_loadu_si512(buffer + idx_decode + 1);
+      __m512i result;
+      if (sizeof(T) == 4) {
+        // 16 float for one m512i block, two bytes in valid_bits
+        __mmask16 k =
+            *(reinterpret_cast<const __mmask16*>(valid_bits + idx_valid_bytes + 1));
+        result = _mm512_mask_expand_epi32(zero, k, src);
+      } else {
+        // 8 double for one m512i block, one byte in valid_bits
+        __mmask8 k = *(valid_bits + idx_valid_bytes + 1);
+        result = _mm512_mask_expand_epi64(zero, k, src);
+      }
+      _mm512_storeu_si512(buffer + idx_spaced + 1, result);
+    } else {  // All null in this batch
+      _mm512_storeu_si512(buffer + idx_spaced + 1, zero);
+    }
+  }
+
+  // The remaining front suffix
+  while (idx_decode < idx_spaced && idx_decode >= 0) {
+    if (BitUtil::GetBit(valid_bits, idx_valid_bits)) {
+      buffer[idx_spaced] = buffer[idx_decode];
+      idx_decode--;
+    }
+    idx_spaced--;
+    idx_valid_bits--;
+  }
+
+  return num_values;
+}
+
+template <typename T>
+inline int PutSpacedAvx512(const T* src, int num_values, const uint8_t* valid_bits,
+                           int64_t valid_bits_offset, T* output) {
+  if (sizeof(T) == 4 || sizeof(T) == 8)  // Based on _mask_compress_epi32/64
+    return PutSpacedAvx512Compress<T>(src, num_values, valid_bits, valid_bits_offset,
+                                      output);
+  else
+    return PutSpacedScalar<T>(src, num_values, valid_bits, valid_bits_offset, output);
+}
+
+template <typename T>
+inline int DecodeSpacedAvx512(T* buffer, int num_values, int null_count,
+                              const uint8_t* valid_bits, int64_t valid_bits_offset) {
+  if (sizeof(T) == 4 || sizeof(T) == 8)  // Based on _mask_expand_epi32/64
```

Comment:
nit: I'm not sure if we've decided on this on the project but I much prefer wrapping single lines in braces even though not required.

Reply:
done